### PR TITLE
    Add MALLOC_CHECK_ and MALLOC_PERTURB_ libc env to the test suite …

### PR DESCRIPTION
**Pull request by devzero2000**
…for detecting heap corruption for glibc <2.34. For glibc >=2.34 use the new heap consistency checking environment variables

    Recent versions of Linux libc (later than 5.4.23) and glibc (2.x)
    include a malloc() implementation which is tunable via environment
    variables. When MALLOC_CHECK_ is set, a special (less efficient)
    implementation is used which is designed to be tolerant against
    simple errors, such as double calls of free() with the same argument,
    or overruns of a single byte (off-by-one bugs). When MALLOC_CHECK_
    is set to 3, a diagnostic message is printed on stderr
    and the program is aborted.

    Setting the MALLOC_PERTURB_ environment variable causes the malloc
    functions in libc to return memory which has been wiped and clear
    memory when it is returned.
    Of course this does not affect calloc which always does clear the memory.

    The reason for this exercise is, of course, to find code which uses
    memory returned by malloc without initializing it and code which uses
    code after it is freed. valgrind can do this but it's costly to run.
    The MALLOC_PERTURB_ exchanges the ability to detect problems in 100%
    of the cases with speed.

    The byte value used to initialize values returned by malloc is the byte
    value of the environment value. The value used to clear memory is the
    bitwise inverse. Setting MALLOC_PERTURB_ to zero disables the feature.

    This technique can find hard to detect bugs.
    It is therefore suggested to always use this flag (at least temporarily)
    when testing out code or a new distribution (http://udrepper.livejournal.com/11429.html).

    We have to be careful if you want to use valgrind with the test suite however.

    Memcheck wraps client calls to malloc(), and puts a "red zone" on
    each end of each block in order to detect access overruns.
    Memcheck already detects double free() (up to the limit of the buffer
    which remembers pending free()). Thus memcheck subsumes all the
    documented coverage of MALLOC_CHECK_.

    If MALLOC_CHECK_ is set non-zero when running memcheck, then the
    overruns that might be detected by MALLOC_CHECK_ would be overruns
    on the wrapped blocks which include the red zones.  Thus MALLOC_CHECK_
    would be checking memcheck, and not the client.  This is not useful,
    and actually is wasteful.  The only possible [documented] advantage
    of using MALLOC_CHECK_ and memcheck together, would be if MALLOC_CHECK_
    detected duplicate free() in more cases than memcheck because memcheck's
    buffer is too small.

    Therefore we shouldn't use use MALLOC_CHECK_ and valgrind at the
    same time.

    MALLOC_CHECK_ and MALLOC_PERTURB_ were the environment variables used until glibc 2.34.
    In glibc >= 2.34 MALLOC_CHECK_ and MALLOC_PERTURB_ environment
    variables have been replaced by GLIBC_TUNABLES.  Also the new
    glibc requires that you preload a library called libc_malloc_debug.so
    to get these features.

    This patch was inspired by a Richard W.M. Jones ndbkit patch
    (https://gitlab.com/nbdkit/nbdkit/-/commit/362e0fdcae37db876e13b944102a5c152e6bc563)

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
@devzero2000 Hey, could you reword the title of your commit so that it fits in <70 characters? It's so long that it spills out and is unreadable when looking through git logs or generating changelogs.

**Comment by devzero2000** [comment]
> @devzero2000 Hey, could you reword the title of your commit so that it fits in <70 characters? It's so long that it spills out and is unreadable when looking through git logs or generating changelogs.

i have just amended the commit message

**Review by Conan-Kudo — APPROVED** [review]

**Comment by pmatilai** [comment]
I'd just drop the pre 2.34 stuff for simplicity, testing is primarily a forward-looking activity.

**Comment by Conan-Kudo** [comment]
I think we try to make sure this stuff works on the latest EL distributions, and neither RHEL nor SLE have releases with glibc 2.34+ right now. So at least for now, I think we should take it as-is.

**Comment by pmatilai** [comment]
Current RHEL and SLES will never use this as they'll have older rpms anyway, they're not the target of rpm upstream development, future is. We care about rpm being compilable there but that's as far as it goes, this is primarily about rpm upstream CI.

Me thinks the right thing here is to just unconditionally set the glibc >= 2.34 magic up for testing, for other platforms its just a no-op and this will cut the patch to 1/3 with no real loss.

**Comment by devzero2000** [comment]
> Current RHEL and SLES will never use this as they'll have older rpms anyway, they're not the target of rpm upstream development, future is. We care about rpm being compilable there but that's as far as it goes, this is primarily about rpm upstream CI.
> 
> Me thinks the right thing here is to just unconditionally set the glibc >= 2.34 magic up for testing, for other platforms its just a no-op and this will cut the patch to 1/3 with no real loss.

I have updated the patch. Thanks

File: tests/atlocal.in [hunk 1]
```diff
@@ -50,6 +50,20 @@ else
     CAP_DISABLED=true;
 fi
 
+# Enable malloc-check as a cheap way to find some use-after-free and
+# uninitialized read problems when using glibc, and doesn't affect
+# normal operation or other libc.
+# Only for GLIBC >=2.34
```
Comment (dmnks) [inline tests/atlocal.in:56]: This seems like an omission from the previous version of this patch that was supposed to check for the GLIBC version.

**Review by dmnks — CHANGES_REQUESTED** [review]
This patch currently fails in CI. Also, see comments inline. Would you be up to revisiting it?

**Comment by pmatilai** [comment]
Superceded by #2218, but thanks for the initiative and the patch!

**Pull request closed without merging**